### PR TITLE
NDArray functions

### PR DESCRIPTION
**Pull request by daniel-goldstein**
- Register arithmetic functions for NDArray w/ NDArray, value w/ NDArray and NDArray w/ value both in Scala and Python
- Add natural number and natural number variable to both Scala and Python and make NDArray's `nDim` a `NatBase`. This allows  specifying function signatures that work on NDArrays of any (but matching) dimensionality.
- Add `NDArrayNumericExpression` and arithmetic methods on it.

### Notes
- In Scala, I directly register the same numeric operations for arrays on ndarrays for reuse and so they don't diverge. However, the IRs generated for `%` and `**` are not supported in C++ yet. I removed the front-end methods for them on `NDArrayNumericExpression` but left the rest of the infrastructure in.

### CODE REVIEW

File: hail/python/hail/expr/expressions/typed_expressions.py [hunk 1]
```diff
@@ -3055,6 +3015,175 @@ def includes_end(self):
         return self._method("includesEnd", tbool)
 
 
+class NDArrayExpression(Expression):
+
```
Comment (tpoterba) [inline hail/python/hail/expr/expressions/typed_expressions.py]: needs docstring

File: hail/python/hail/expr/expressions/typed_expressions.py [hunk 2]
```diff
@@ -3055,6 +3015,175 @@ def includes_end(self):
...
+    @property
+    def ndim(self):
```
Comment (tpoterba) [inline hail/python/hail/expr/expressions/typed_expressions.py]: needs docstring

File: hail/python/hail/expr/expressions/typed_expressions.py [hunk 3]
```diff
@@ -3055,6 +3015,175 @@ def includes_end(self):
...
+        """
+        return self._bin_op_numeric("*", other)
+
+    def __rmul__(self, other):
+        return self._bin_op_numeric_reverse("*", other)
+
+    @staticmethod
+    def _div_ret_type_f(t):
+        assert is_numeric(t)
+        if t == tint32 or t == tint64:
```
Comment (tpoterba) [inline hail/python/hail/expr/expressions/typed_expressions.py]: missing bool.

File: hail/python/hail/expr/type_parsing.py [hunk 4]
```diff
@@ -14,9 +14,10 @@
     bool = "tbool" / "bool"
     call = "tcall" / "call"
     str = "tstr" / "str"
+    nat = ("tnat" / "nat") _ "<" identifier ">"
     locus = ("tlocus" / "locus") _ "<" identifier ">"
     array = ("tarray" / "array") _ "<" type ">"
-    ndarray = ("tndarray" / "ndarray") _ "<" type "," identifier ">"
+    ndarray = ("tndarray" / "ndarray") _ "<" type "," type ">"
```
Comment (tpoterba) [inline hail/python/hail/expr/type_parsing.py]: the type here should be `nat`, not `type`. Identifier was also wrong.

File: hail/python/hail/expr/type_parsing.py [hunk 5]
```diff
@@ -14,9 +14,10 @@
     bool = "tbool" / "bool"
     call = "tcall" / "call"
     str = "tstr" / "str"
+    nat = ("tnat" / "nat") _ "<" identifier ">"
```
Comment (tpoterba) [inline hail/python/hail/expr/type_parsing.py]: this shouldn't be an identifier, it should be an integer literal. no?

File: hail/python/hail/expr/types.py [hunk 6]
```diff
@@ -519,7 +520,7 @@ class tndarray(HailType):
     :class:`.NDArrayExpression`, :func:`.ndarray`
     """
 
-    @typecheck_method(element_type=hail_type, ndim=int)
+    @typecheck_method(element_type=hail_type, ndim=hail_type)
     def __init__(self, element_type, ndim):
```
Comment (tpoterba) [inline hail/python/hail/expr/types.py]: I'm a bit wary about exposing `ndim` to the user as a type, rather than a number. Can we discuss this tomorrow?

File: hail/src/main/scala/is/hail/cxx/Emit.scala [hunk 7]
```diff
@@ -861,6 +861,7 @@ class Emitter(fb: FunctionBuilder, nSpecialArgs: Int, ctx: SparkFunctionContext)
       case ir.NDArrayMap(child, elemName, body) =>
         val childTyp = child.pType.asInstanceOf[PNDArray]
         val elemPType = childTyp.elementType
+        val nDims = childTyp.nDims.asInstanceOf[PNat].n
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/cxx/Emit.scala]: `Nat` definitely doesn't need a physical type.

File: hail/python/hail/expr/type_parsing.py (hunk 5 above)
Comment (daniel-goldstein) [reply hail/python/hail/expr/type_parsing.py]: I can add another parsing rule that explicitly only matches on integers.

File: hail/python/hail/expr/type_parsing.py (hunk 4 above)
Comment (daniel-goldstein) [reply hail/python/hail/expr/type_parsing.py]: `tvariable` is also an option, but I can restrict it to only those two.

File: hail/python/hail/expr/types.py (hunk 6 above)
Comment (daniel-goldstein) [reply hail/python/hail/expr/types.py]: Ya, definitely. I think it's preferable that it be internal, but it introduces a bit of complexity in that you have to be careful to chose between `nd_expr.ndim` and `nd_expr._type.ndim`.

File: hail/src/main/scala/is/hail/cxx/Emit.scala (hunk 7 above)
Comment (daniel-goldstein) [reply hail/src/main/scala/is/hail/cxx/Emit.scala]: Cool, should `TNat` throw an unsupported exception for its `physicalType`?

File: hail/python/hail/expr/types.py [hunk 8]
```diff
@@ -565,18 +567,30 @@ def _pretty(self, l, indent, increment):
         l.append('ndarray<')
         self.element_type._pretty(l, indent, increment)
         l.append(', ')
-        l.append(str(self.ndim))
+        self.element_type._pretty(l, indent, increment)
```
Comment (tpoterba) [inline hail/python/hail/expr/types.py]: bug

Comment (tpoterba) [reply hail/python/hail/expr/types.py]: and should be just integer

File: hail/python/hail/expr/types.py [hunk 9]
```diff
@@ -542,7 +543,7 @@ def ndim(self):
 
         Returns
         -------
-        `int32`
+        :class:`.HailType`
```
Comment (tpoterba) [inline hail/python/hail/expr/types.py]: should be integer, with checking for tnat

File: hail/python/hail/expr/type_parsing.py (hunk 5 above)
Comment (tpoterba) [reply hail/python/hail/expr/type_parsing.py]: also add rule that accepts integers as nats

Comment (tpoterba) [reply hail/python/hail/expr/type_parsing.py]: actually accepting either a type or a integer literal in the ndarray last argument seems better

Comment (tpoterba) [reply hail/python/hail/expr/type_parsing.py]: (add a new rule for int literal and a visitor which returns the int)

**Comment by daniel-goldstein** [comment]
nats are now not types and align more with the ReferenceGenome structure, making for better function signatures. There's a `NatBase` with `Nat` and `NatVariable` as extending classes. For `NatVariable`, instances mutate a class variable `nat`. Type variables behave similar but using a map of name -> box, but I didn't really see the point of ever needing more than one `nat` variable in the same context. Either way it seems a little weird so glad to take feedback.

**Comment by daniel-goldstein** [comment]
Also updated binary ops to promote arg types to the return type (e.g. int32 -> float64) if the return type is also numeric (not bools though). This eliminates the need to double-check the signature of a division both in the IR and expressions.

File: hail/python/hail/expr/expressions/base_expression.py [hunk 10]
```diff
@@ -388,10 +388,17 @@ def _promote_numeric(self, typ):
         coercer = expressions.coercer_from_dtype(typ)
         if isinstance(typ, tarray) and not isinstance(self.dtype, tarray):
             return coercer.ec.coerce(self)
+        elif isinstance(typ, tndarray) and not isinstance(self.dtype, tndarray):
+            return coercer.ec.coerce(self)
         else:
             return coercer.coerce(self)
 
-    def _bin_op_numeric_unify_types(self, name, other):
+    @staticmethod
```
Comment (tpoterba) [inline hail/python/hail/expr/expressions/base_expression.py]: I think these changes need to be in a separate PR - they need a deeper review. Is that a lot of trouble?

Comment (daniel-goldstein) [reply hail/python/hail/expr/expressions/base_expression.py]: np I can move them

Comment (daniel-goldstein) [reply hail/python/hail/expr/expressions/base_expression.py]: Should be good now

File: hail/python/hail/expr/types.py [hunk 11]
```diff
@@ -1449,7 +1463,7 @@ class tvariable(HailType):
         'float64': lambda x: x == tfloat64,
         'locus': lambda x: isinstance(x, tlocus),
         'struct': lambda x: isinstance(x, tstruct),
-        'tuple': lambda x: isinstance(x, ttuple)
+        'tuple': lambda x: isinstance(x, ttuple),
```
Comment (tpoterba) [inline hail/python/hail/expr/types.py]: unnecessary change

File: hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala [hunk 12]
```diff
@@ -8,6 +8,19 @@ import is.hail.expr.types.virtual._
 import is.hail.utils._
 
 object ArrayFunctions extends RegistryFunctions {
+  val arrayOps: Array[(String, Type, Type, (IR, IR) => IR)] =
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala]: move back? This didn't change, right?

Comment (daniel-goldstein) [reply hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala]: I moved this out of the `registerAll` method so I could register the same functions for NDArrays.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala]: ah, got it.

File: hail/python/hail/expr/types.py (hunk 11 above)
Comment (daniel-goldstein) [reply hail/python/hail/expr/types.py]: removed

File: hail/python/hail/expr/expressions/typed_expressions.py [hunk 13]
```diff
@@ -3055,6 +2998,188 @@ def includes_end(self):
...
+        >>> names.ndim
+        2
+
+        Returns
+        -------
+        :obj:`int`
+        """
+        return self._type.ndim
+
+    @typecheck_method(item=oneof(expr_int64, tupleof(expr_int64)))
```
Comment (tpoterba) [inline hail/python/hail/expr/expressions/typed_expressions.py]: should definitely support slicing at some point soon too.

File: hail/python/hail/expr/expressions/typed_expressions.py [hunk 14]
```diff
@@ -3055,6 +2998,188 @@ def includes_end(self):
...
+
+        if len(item) != self.ndim:
+            raise ValueError(f'Must specify one index per dimension. '
+                             f'Expected {self.ndim} dimensions but got {len(item)}')
+
+        # indexes must be iterable
+        if len(item) == 0:
+            idxs = hl.empty_array(hl.tint64)
+        else:
+            idxs = hl.array(list(item))
```
Comment (tpoterba) [inline hail/python/hail/expr/expressions/typed_expressions.py]: I don't totally understand why the index is an array IR, instead of a sequence of integers. This also makes it very difficult to build slicing on top. Let's discuss tomorrow.
